[Feature] Metadata editing for JPG images

Spoicy opened the pull request:
This PR addresses an old feature request (issue #139) regarding metadata editing.

This pull request does the following:
- Adds a new service Metadata
- Migrates usage of metadata functions from IMGTools to Metadata
- Adds Composer package [FileEye/PEL](https://github.com/FileEye/pel)
- Adds editing support to EXIF and IPTC Metadata for JPG images
- Reworks initial metadata reading to work with PEL
- Adds new tab in Image form for editing metadata via subforms
- Includes button in Image form to save metadata to an image

### How could this effect performance?

The composer package is loaded in via the providers.php file, which should not effect performance in any noticeable way. In order to edit EXIF metadata, [PEL](https://github.com/FileEye/pel) is used instead of [exif_read_data()](https://www.php.net/manual/en/function.exif-read-data.php), as the naming of metadata fields can be different. This has a slight effect on the performance when uploading images.

### How to test this PR?

- Install the zip from the [PR Build Artifact](https://github.com/JoomGalleryfriends/JoomGallery/actions/runs/15750896549/job/44395747715?pr=207) workflow run
- Prepare a JPG file locally with EXIF and IPTC metadata
- Upload the prepared JPG image
- Edit the different metadata fields and save the Image form to store them in the database.
- Save the metadata to the image via the "Save metadata to file" button.
- Download the Original file of the edited image and check if the new metadata was saved.
- Test also the "Replace with metadata" feature

### Possible Software to edit and view metadata locally
- [Exif Pilot](https://www.colorpilot.com/exif.html)
- [XnView](https://www.xnview.com/de/xnviewmp/#downloads)

Elfangor93 commented on administrator/com_joomgallery/sql/updates/mysql/4.0.0.sql:
```diff
@@ -133,6 +133,7 @@ CREATE TABLE IF NOT EXISTS `#__joomgallery_configs` (
 `jg_imgprocessor` VARCHAR(5) NOT NULL DEFAULT "gd",
 `jg_fastgd2creation` TINYINT(1) NOT NULL DEFAULT 1,
 `jg_impath` VARCHAR(100) NOT NULL DEFAULT "",
+`jg_metaprocessor` VARCHAR(15) NOT NULL DEFAULT "php",
```
This has to go into the file 4.1.0.sql

Spoicy replied:
I've moved it out of 4.0.0.sql and added it into 4.1.0.sql as an ALTER TABLE statement.

Elfangor93 commented on administrator/com_joomgallery/forms/config.xml:
```diff
@@ -253,6 +253,15 @@
+             label="COM_JOOMGALLERY_CONFIG_METAPROCESSOR"
+             description="COM_JOOMGALLERY_CONFIG_METAPROCESSOR_LONG" >
+         <option value="php">PHP</option>
+         <option value="exiftools">COM_JOOMGALLERY_CONFIG_METADATA_EXIFTOOLS</option>
```
As the exif-tools part of the metadata service is not yet existing, I would remove this as an option in the config.

Spoicy replied:
Removed the exiftools option as requested, the rest of the config remains.

Elfangor93 commented on administrator/com_joomgallery/src/Service/Metadata/MetadataInterface.php:
```diff
@@ -0,0 +1,78 @@
+interface MetadataInterface
+{
+  /**
+   * 
```
Method description comments missing.

Elfangor93 commented on composer.lock:
I would add this file to the .gitignore and remove it from the repository.

Spoicy replied:
Added to the .gitignore and removed from the repository.

Elfangor93 reviewed (dismissed):
Great pull request!
Thank you very much. Just scrolled through the code. Looks nice and clean.

Elfangor93 commented on administrator/com_joomgallery/src/Service/Uploader/Uploader.php:
```diff
@@ -244,8 +244,11 @@ public function overrideData(&$data): bool
     // Create the IMGtools service
```
When I see this correct, then the IMGtools service is not needed anymore in this method Uploader::overrideData().
Therefore createIMGtools() and delIMGtools() could be deleted in this method.

Elfangor93 commented on administrator/com_joomgallery/src/Service/Metadata/IptcDataEditor.php:
```diff
@@ -0,0 +1,140 @@
+    {
+        $retval = "";
+        foreach ($app13 as $tag => $value) {
+            var_dump($value);
```
I guess this var_dump() is not really needed ;-)

Spoicy replied:
I removed it ;-)

Elfangor93 commented on administrator/com_joomgallery/src/Controller/ImageController.php:
```diff
@@ -266,4 +266,56 @@ public function cancel($key = null)
+   *
+   * @return  boolean  True if metada was saved successfully, false otherwise.
+   *
+   * @since   4.0
```
New added classes should get a `@since 4.1.0` as we merge this PR into v4.1.0

Spoicy replied:
Adjusted the comments to be ```@since 4.1.0```

Elfangor93 commented on administrator/com_joomgallery/src/Service/Metadata/MetadataInterface.php:
```diff
@@ -0,0 +1,84 @@
+   * 
+   * @return mixed                Image data to be stored with Filemanager
+   */
+  public function writeMetadata($img, $imgmetadata, $local_source);
```
Please update the Interface based on what you coded in MetadataPHP. The argument `$local_source` for example should have a default value.

Spoicy replied:
Updated the argument to have a default value of true to match the implementation.

Elfangor93 commented on administrator/com_joomgallery/src/Service/Metadata/MetadataPHP.php:
```diff
@@ -0,0 +1,600 @@
+   * 
+   * @since 4.0.0
+   */
+  private function getPelImageObjects(string $img)
```
Why have you chosen them to have the `private` scope and not the `protected` scope?
If you dont have an explicit reason to use `private`, please use `protected`.

Elfangor93 commented on administrator/com_joomgallery/layouts/joomla/form/field/subform/exif.php:
```diff
@@ -0,0 +1,47 @@
+
+/**
+ ******************************************************************************************
+ **   @version    4.0.0-dev                                                                  **
```
Please update the file header to our new file header without the `@version` line.

Spoicy replied:
Updated all of the Metadata files to use the new header

Elfangor93 commented:
@Spoicy This line in the file `administrator/com_joomgallery/tmpl/image/edit.php` can be removed. Otherwise, we have the metadata form twice.

```
<?php echo $this->form->renderField('imgmetadata'); ?>
```

Elfangor93 commented on administrator/com_joomgallery/tmpl/image/edit.php:
```diff
@@ -171,6 +171,19 @@
+
+  <?php echo HTMLHelper::_('uitab.addTab', 'myTab', 'Metadata', Text::_('COM_JOOMGALLERY_METADATA', true)); ?>
+  <div class="row">
+    <div class="col-6">
```
Please change this class to `col-12` in order to have the form full width.

Elfangor93 commented:
The mapping of the IPTC metadata fields is not matching the mapping in other image tools. I think for user convenience, this mapping should be fitting with as much image tools as possible.

[Exif Pilot](https://www.colorpilot.com/exif.html)
![grafik](https://github.com/user-attachments/assets/76022a48-2699-48b6-a0ef-5d9a1a7e2a90)

[XnView](https://www.xnview.com/de/)
![grafik](https://github.com/user-attachments/assets/d63bc4bc-1595-4a00-8956-ea77a2569610)

JoomGallery with this PR
![grafik](https://github.com/user-attachments/assets/9c19a9c6-1f71-42e1-8e55-b5784fbe2aa8)

Elfangor93 commented:
The comment field is also not properly filled out.

## XnView
![grafik](https://github.com/user-attachments/assets/aed8d005-ff10-455b-84aa-8cb01a80685f)

## JoomGallery with this PR applied
![grafik](https://github.com/user-attachments/assets/3e067b49-fbe7-4536-bea4-bd9f3ad5856c)

MrMusic commented:
>Only discussion point here are the fields Byline and Byline Title. To be honest the term Byline is very confusing to me...

I would recommend using the official names published on [iptc.org](https://www.iptc.org/std/photometadata/documentation/userguide/#field-reference-table) if possible.
In the case of `Byline`, you could add an addition in brackets, e.g: 'Byline (Creator)'

>Maybe the comment field is better placed outside of the EXIF form, but could be placed on the very top, above the IPTC data.

In my opinion, the comment field is independent of Exif and IPTC. That's why I would display it independently.

PS: Can this PR already be tested by downloading the zip from Github like the other PRs?

Elfangor93 commented:
> Can this PR already be tested by downloading the zip from Github like the other PRs?

No. And it probably never will. You eighter have to add PEL using composer or by using the additional zip file provided.

MrMusic commented:
>If you dont wan't or you just can't run this command, you can alternatively download this zip and copy the containing folder in your Joomla! installation folder after you installed the PR.

Nevertheless, with or without the contents of administrator.zip, the PR cannot be installed:
```
JInstaller: :Install: File does not exist [ROOT]\tmp\install_68262171dad40\JoomGallery-project-metadata\administrator\com_joomgallery\vendor
Component Update: Failed to copy administrator files.
```
It looks to me as if the `<folder>vendor</folder>` in joomgallery.xml is wrong.

Elfangor93 commented:
> It looks to me as if the <folder>vendor</folder> in joomgallery.xml is wrong.

The` joomgallery.xml` is correct. The zip file of the Spoicy:project-metadata is just missing this vendor folder as it is added with composer during release. But I dont want to mess up the Repo with 3rd party libraries.

I added a description in the first comment on how to adjust the zip file if you want to install the PR using the Joomla Installer.

MrMusic commented:
>No. And it probably never will. You eighter have to add PEL using composer or by using the additional zip file provided.

I'm not sure whether the advantages outweigh the disadvantages.

1. To test a PR the hurdle for the testers is unfortunately higher.
Until now it was relatively easy to test a PR because you could download the zip from Github and install it. This will no longer be possible in future.
To test a PR in the future, the hurdle for the testers will be higher.

2. For the final release, `PEL` must still be included in the zip.
 This means that we are still the 'distributor' and are responsible for delivering a secure and error-free version.

Spoicy commented:
Implemented changes for the two notable problems after discussion with @Elfangor93:

Added the comment property back into imgmetadata, this was a slight misunderstanding on my part of what it represented in ```exif_read_data()```. This is done with that method like before, which to me personally is a bit unfortunate to read the exif data twice. If desired, the code could be reworked to only use ```exif_read_data()``` and shouldn't be too tricky, however the naming of the fields between PEL and ```exif_read_data()``` needs to be guaranteed.

Fixed the issue with IPTC keywords, the value now gets converted to a string for the db, and converted back into an array when writing into the image.

Elfangor93 commented:
During testing, I have seen that the Keywords are not correctly threated.

Initially, I set the Keywords in Exif Pilot as following:
![grafik](https://github.com/user-attachments/assets/6b8d0c0b-3b96-471f-b617-c0ae0badac25)

Then I uploaded the image to JoomGallery and it is shown like this:
![grafik](https://github.com/user-attachments/assets/c619bdd7-669b-4407-9ea7-27d90f43748c)

Afterwards I saved the metadata to file and downloaded the image and viewed it again in Exif Pilot. The keywords are now shown like this:
![grafik](https://github.com/user-attachments/assets/4262eb5a-a273-4313-82f6-b3fd73442f7c)

I guess the Keywords are corrupted when reading them out from the metadata.

Elfangor93 commented:
In multiple exif form fields it is not possible to enter special characters like "ö" or "ü". But in Exif Pilot I am able to enter this characters.
![grafik](https://github.com/user-attachments/assets/7ec4f95a-3ddf-43da-bd5e-a70c87ba0c1a)
I guess this is due to the pattern attribute in the 
![grafik](https://github.com/user-attachments/assets/0dd67314-4dfe-4ece-9c5c-81d9b2f0f567)

Elfangor93 commented:
If you create a new image (single upload) and you fill out something in the tab "metadata" your entry is ignored and the metadata of the image is taken instead.
And if you change any of the dropdown fields in the tab "metadata" before uploading, the upload will fail.

My proposition:
Show the tab metadata only when the form is not empty - show thet ab metadata only when the image is created/uploaded.

MrMusic commented:
If an uploaded image does not contain any Exif data, during saving 'default values' are still saved in the metadata field.
Example: Metering Mode, Light Source, Flash...
My opinion: If an image had no Exif data, nothing should be saved automatically in the metadata field.

eumel1602 commented:
Is it possible that no metadata is being saved with the current download?
I've tried everything for half an hour now. No changes to the data point are visible when viewing the image in the JG...

Spoicy commented:
Regarding the special characters, the class used to read/write those fields is PelEntryAscii, which is why I implemented a regex to disallow non-ASCII characters. After some testing, PEL doesn't actually check for this however. I've since removed the regex pattern from those fields as I was requested to do.

Keywords should save correctly now if written by Exif Pilot, the program writes additional \u0000 characters after each keywords, which now get removed when reading from the image.

Elfangor93 commented:
> If an uploaded image does not contain any Exif data, during saving 'default values' are still saved in the metadata field.

Its fixed now. There is now a default field value which is empty:
![grafik](https://github.com/user-attachments/assets/9cf0d9ff-2b0c-4abb-9b96-ffaeca6aab2e)
![grafik](https://github.com/user-attachments/assets/4e5942fd-adf4-4c27-9e8a-645f84754d80)

Elfangor93 commented:
I have tested this item ✅ successfully. Thank you very much!

Please @all: test this PR. I really would like to merge this by tomorrow (01.07.2025).

Elfangor93 commented:
I have tested this item ✅ successfully. Thank you very much!

Please @all test this PR. I really would like to merge this by tomorrow (01.07.2025).

MrMusic commented:
>GPS Infos were not yet implemented. I can add this in today, would the required fields be the three listed in the translation file (altitude, longitude, latitude)?

Irfanview shows the GPS data like this:
![gpsdata-irfanview](https://github.com/user-attachments/assets/8e00c98c-672c-4fe8-ba1d-65f764dcf6e7)

Geosetter (Exif-tool) show the GPS data like this:
![gpsdata-geosetter](https://github.com/user-attachments/assets/1cf5714d-8d8f-4d67-84a5-0dd6ce291a8b)

>GPS data from EXIF was to my knowledge not yet supported by JoomGallery, it only took the IFD0 and EXIF entries from exif_read_data, which is why GPS is not implemented in this PR either.

JoomGallery 3 can show the GPS data (and a google map too). So I think the GPS data are important.
Maybe helpful: https://reintech.io/blog/comprehensive-guide-php-exif-library-image-metadata

Elfangor93 commented:
@Spoicy Whats your impression, how long does it take to implement GPS Exif data?
Otherwise we could also merge this PR and add the GPS Exif data in a later PR and merge it for v4.2?

I really want to release the v4.1 very soon...

Elfangor93 commented:
Good, then we do this. Thanks @Spoicy for your great contribution!
